File: templates/scenarios/react-vite/api/internal/testutil/modeltest/formal_test.go
package modeltest_test

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"
	"testing"
	"{{SCENARIO_ID}}/internal/testutil/modeltest"

	"github.com/stretchr/testify/require"
)

func TestValidateFormalArtifactFresh_AcceptsCurrentModelHash(t *testing.T) {
	modelPath := writeFormalModel(t, "module Example {}\n")
	contractPath := writeFormalModel(t, "{}\n")
	generatorPath := writeFormalModel(t, "#!/usr/bin/env node\n")
	artifact := validFormalArtifact(t, contractPath, modelPath, generatorPath)

	errs := modeltest.ValidateFormalArtifactFresh(artifact, formalExpectation(contractPath, modelPath, generatorPath))
	require.Empty(t, errs)
}

func TestValidateFormalArtifactFresh_RejectsStaleHashAndMissingChecks(t *testing.T) {
	modelPath := writeFormalModel(t, "module Example {}\n")
	contractPath := writeFormalModel(t, "{}\n")
	generatorPath := writeFormalModel(t, "#!/usr/bin/env node\n")
	artifact := validFormalArtifact(t, contractPath, modelPath, generatorPath)
	artifact.Source.ModelSHA256 = "0000000000000000000000000000000000000000000000000000000000000000"
	artifact.Checks.Verified = false

	errs := modeltest.ValidateFormalArtifactFresh(artifact, formalExpectation(contractPath, modelPath, generatorPath))
	require.NotEmpty(t, errs)
	require.Contains(t, joined(errs), "formal artifact modelSha256=")
	require.Contains(t, joined(errs), "formal artifact was not verified")
}

func TestValidateFormalTransitionsReplay_AcceptsGeneratedMatrix(t *testing.T) {
	artifact := validFormalArtifact(t, writeFormalModel(t, "{}\n"), writeFormalModel(t, "module Example {}\n"), writeFormalModel(t, "#!/usr/bin/env node\n"))

	errs := modeltest.ValidateFormalTransitionsReplay(
		artifact,
		[]status{statusIdle, statusBusy, statusDone},
		[]event{eventStart, eventFinish},
		transition,
	)
	require.Empty(t, errs)
}

func TestValidateFormalTransitionsReplay_RejectsUnknownAndDivergentTransition(t *testing.T) {
	artifact := validFormalArtifact(t, writeFormalModel(t, "{}\n"), writeFormalModel(t, "module Example {}\n"), writeFormalModel(t, "#!/usr/bin/env node\n"))
	artifact.Transitions[0].To = "done"
	artifact.Transitions[1].Event = "ghost"

	errs := modeltest.ValidateFormalTransitionsReplay(
		artifact,
		[]status{statusIdle, statusBusy, statusDone},
		[]event{eventStart, eventFinish},
		transition,
	)
	require.NotEmpty(t, errs)
	require.Contains(t, joined(errs), "unknown event ghost")
}

func TestValidateFormalTracesReplay_RejectsUnknownAndDivergentTrace(t *testing.T) {
	artifact := validFormalArtifact(t, writeFormalModel(t, "{}\n"), writeFormalModel(t, "module Example {}\n"), writeFormalModel(t, "#!/usr/bin/env node\n"))
	artifact.NamedTraces[0].Steps[0].Want = "done"
	artifact.NamedTraces[0].Steps[1].Event = "ghost"

	errs := modeltest.ValidateFormalTracesReplay(
		artifact,
		[]status{statusIdle, statusBusy, statusDone},
		[]event{eventStart, eventFinish},
		transition,
	)
	require.NotEmpty(t, errs)
	require.Contains(t, joined(errs), "unknown event ghost")
}

func formalExpectation(contractPath string, modelPath string, generatorPath string) modeltest.FormalArtifactExpectation {
	return modeltest.FormalArtifactExpectation{
		ContractPath:  contractPath,
		ModelPath:     modelPath,
		GeneratorPath: generatorPath,
		Invariants:    []string{"TypeOK", "TerminalClosure"},
	}
}

func validFormalArtifact(t *testing.T, contractPath string, modelPath string, generatorPath string) modeltest.FormalArtifact {
	t.Helper()
	return modeltest.FormalArtifact{
		SchemaVersion: 2,
		FlowID:        "example.flow",
		Source: modeltest.FormalArtifactSource{
			ContractPath:        contractPath,
			ContractSHA256:      fileSHA256(t, contractPath),
			GeneratorPath:       generatorPath,
			GeneratorSHA256:     fileSHA256(t, generatorPath),
			GeneratorVersion:    2,
			ModelPath:           modelPath,
			ModelSHA256:         fileSHA256(t, modelPath),
			QuintVersion:        "0.32.0",
			VerificationBackend: "apalache",
		},
		Commands: map[string][]string{
			"typecheck": {"quint", "typecheck", modelPath},
			"test":      {"quint", "test", modelPath},
			"verify":    {"quint", "verify", modelPath},
			"run":       {"quint", "run", modelPath},
		},
		States: []string{"idle", "busy", "done"},
		Events: []string{"start", "finish"},
		Transitions: []modeltest.FormalArtifactTransition{
			{From: "idle", Event: "start", To: "busy", WantError: false},
			{From: "idle", Event: "finish", To: "idle", WantError: true},
			{From: "busy", Event: "start", To: "busy", WantError: true},
			{From: "busy", Event: "finish", To: "done", WantError: false},
			{From: "done", Event: "start", To: "done", WantError: true},
			{From: "done", Event: "finish", To: "done", WantError: true},
		},
		NamedTraces: []modeltest.FormalArtifactTrace{
			{
				Name:    "generated",
				Initial: "idle",
				Steps: []modeltest.FormalArtifactTraceStep{
					{Event: "start", Want: "busy", WantError: false},
					{Event: "finish", Want: "done", WantError: false},
				},
			},
		},
		GeneratedTraces: []modeltest.FormalArtifactTrace{
			{
				Name:    "generated_model_001",
				Initial: "idle",
				Steps: []modeltest.FormalArtifactTraceStep{
					{Event: "start", Want: "busy", WantError: false},
				},
			},
		},
		Invariants: []string{"TypeOK", "TerminalClosure"},
		Coverage: modeltest.FormalArtifactCoverage{
			AllStatesCovered:      true,
			AllEventsCovered:      true,
			AllPairsCovered:       true,
			TerminalStatesChecked: true,
		},
		Checks: modeltest.FormalArtifactChecks{
			Typechecked:           true,
			Tested:                true,
			Verified:              true,
			GeneratedFromContract: true,
			GeneratedFromModel:    true,
		},
	}
}

func writeFormalModel(t *testing.T, content string) string {
	t.Helper()
	path := filepath.Join(t.TempDir(), "model.qnt")
	require.NoError(t, os.WriteFile(path, []byte(content), 0o600))
	return path
}

func fileSHA256(t *testing.T, path string) string {
	t.Helper()
	data, err := os.ReadFile(path)
	require.NoError(t, err)
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

File: templates/scenarios/react-vite/ui/src/test-utils/modeltest/formal.test.ts
import { describe, expect, it } from "vitest";

import {
  validateFormalArtifactFresh,
  validateFormalTransitionsReplay,
  validateFormalTracesReplay,
  type FormalArtifact,
} from "./formal";

type Status = "idle" | "busy" | "done";
type Event = "start" | "finish";

const transition = (status: Status, event: Event): Status => {
  if (status === "idle" && event === "start") {
    return "busy";
  }
  if (status === "busy" && event === "finish") {
    return "done";
  }
  throw new Error("illegal transition");
};

const statuses = ["idle", "busy", "done"] as const;
const events = ["start", "finish"] as const;
const modelPath = "ui/src/features/example/model.qnt";
const modelSha256 = "a".repeat(64);

describe("formal modeltest helpers", () => {
  it("accepts a fresh artifact", () => {
    expect(validateFormalArtifactFresh(validArtifact(), { modelPath, modelSha256 })).toEqual([]);
  });

  it("rejects stale hashes and missing checks", () => {
    const artifact = {
      ...validArtifact(),
      source: { ...validArtifact().source, modelSha256: "0".repeat(64) },
      checks: { ...validArtifact().checks, verified: false },
    };

    const errors = validateFormalArtifactFresh(artifact, { modelPath, modelSha256 });
    expect(errors.join("\n")).toContain("formal artifact modelSha256=");
    expect(errors.join("\n")).toContain("formal artifact was not verified");
  });

  it("replays generated transitions", () => {
    const artifact = validArtifact();
    expect(validateFormalTransitionsReplay(artifact, statuses, events, transition)).toEqual([]);
  });

  it("rejects unknown and divergent transitions", () => {
    const artifact = validArtifact();
    const transitions = artifact.transitions.map((transitionRow, index) => {
      if (index === 0) {
        return { ...transitionRow, to: "done" };
      }
      if (index === 1) {
        return { ...transitionRow, event: "ghost" };
      }
      return transitionRow;
    });

    const errors = validateFormalTransitionsReplay({ ...artifact, transitions }, statuses, events, transition);
    expect(errors.join("\n")).toContain("unknown event ghost");
  });

  it("rejects unknown and divergent traces", () => {
    const artifact = validArtifact();
    const traces = [
      {
        name: artifact.traces[0]?.name ?? "",
        initial: artifact.traces[0]?.initial ?? "",
        steps: (artifact.traces[0]?.steps ?? []).map((step, index) => {
          if (index === 0) {
            return { ...step, want: "done" };
          }
          if (index === 1) {
            return { ...step, event: "ghost" };
          }
          return step;
        }),
      },
    ];

    const errors = validateFormalTracesReplay({ ...artifact, traces }, statuses, events, transition);
    expect(errors.join("\n")).toContain("unknown event ghost");
  });
});

const validArtifact = (): FormalArtifact => ({
  schemaVersion: 1,
  flowId: "example.flow",
  source: {
    modelPath,
    modelSha256,
    quintVersion: "0.32.0",
  },
  commands: {
    typecheck: ["quint", "typecheck", modelPath],
    test: ["quint", "test", modelPath],
    verify: ["quint", "verify", modelPath],
    run: ["quint", "run", modelPath],
  },
  states: ["idle", "busy", "done"],
  events: ["start", "finish"],
  transitions: [
    { from: "idle", event: "start", to: "busy", wantError: false },
    { from: "idle", event: "finish", to: "idle", wantError: true },
    { from: "busy", event: "start", to: "busy", wantError: true },
    { from: "busy", event: "finish", to: "done", wantError: false },
    { from: "done", event: "start", to: "done", wantError: true },
    { from: "done", event: "finish", to: "done", wantError: true },
  ],
  traces: [
    {
      name: "generated",
      initial: "idle",
      steps: [
        { event: "start", want: "busy", wantError: false },
        { event: "finish", want: "done", wantError: false },
      ],
    },
  ],
  checks: {
    typechecked: true,
    tested: true,
    verified: true,
    generatedFromModel: true,
  },
});
